verilog/analysis/flow_tree.h

```diff
@@ -0,0 +1,158 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_FLOW_TREE_H_
+#define VERIBLE_VERILOG_FLOW_TREE_H_
+
+#include <bitset>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "absl/status/status.h"
+#include "common/lexer/token_stream_adapter.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+
+// FlowTree class builds the control flow graph of a tokenized System-Verilog
+// source code. Furthermore, enabling doing the following queries on the graph:
+// - Generating all the possible variants (provided via a callback function).
+class FlowTree {
+ private:
+  // 'kMaxDistinctMacros' shows the maximum number of distinct macros that can
+  // be considered in conditonal directives.
+  static constexpr int kMaxDistinctMacros = 128;
+
+ public:
+  using BitSet = std::bitset<kMaxDistinctMacros>;
+  using TokenSequenceConstIterator = verible::TokenSequence::const_iterator;
+
+  // "ConditionalBlock" saves locations of conditionals in a "TokenSequence".
+  //  All locations should point inside this specific "TokenSequence".
+  //  Since it is only used in conjunction with a "TokenSequence",
+  //  It should be initialized to last location in this "TokenSequence",
+  //  For example in "GenerateControlFlowTree" is initialized to
+  //  'source_sequence_.end()'.
+
+  struct Variant {
+    // Contains the token sequence of the variant.
+    verible::TokenSequence sequence;
+
+    // The i-th bit in "macros_mask" is 1 when the macro (with ID = i) is
+    // assumed to be defined, otherwise it is assumed to be undefined.
+    BitSet macros_mask;
+
+    // The i-th bit in "visited" is 1 when the macro (with ID = i) is visited or
+    // assumed (either defined or not), otherwise it is not visited (its value
+    // doesn't affect this variant).
+    //
+    // e.g.:
+    //  `ifdef A
+    //    `ifdef B
+    //    ...
+    //    `endif
+    //  `endif
+    //
+    // Consider the variant in which A is undefined,
+    // we notice that B doesn't affect the variant.
+    // Then the bit corresponding to B in "visited" is 0.
+    BitSet visited;
+  };
+
+  // Receive a complete token sequence of one variant.
+  // variant_sequence: the generated variant token sequence.
+  using VariantReceiver = std::function<bool(const Variant &variant)>;
+
+  explicit FlowTree(verible::TokenSequence source_sequence)
+      : source_sequence_(std::move(source_sequence)){};
+
+  // Generates all possible variants.
+  absl::Status GenerateVariants(const VariantReceiver &receiver);
```

Comment:
Not for now, but one TODO for future work would be to make this a const function
```
absl::Status GenerateVariants(const VariantReceiver &receiver) const;
```

For that to work, we would have to call `GenerateControlFlowTree()` outside as it modifies the state. You previously had that, but that was at a state when it was not quite clear yet what the final API looks like, so we can revisit.
We discussed before that it might be slightly confusing in the API to have such call, when it is called `GenerateControlFlowTree()` as it might not be quite clear to a reader of the API if it is required.

So we can consider these options: 
   * Do it in the constructor. But it would somewhat violate the [don't do work in the constructor](https://google.github.io/styleguide/cppguide.html#Doing_Work_in_Constructors) guidance, in particular since there can be error conditions that we can't return in the constructor.
   * Require the user to call `GenerateControlFlowTree()`, but since that is unnecessarily obscure to the potential usr of the API, we could just call it `Initialize()` ? That way, for someone reading the API, they know what to do.
   * If the user has not called `Initialize()` have `GenerateVariants()` return a status error, maybe `absl::FailedPreconditionError()`. If Initialize() was called, but had some non-ok status, we can return that. So in general, we can have an `absl::Status initialize_status_` in the `FlowTree` class, that we initialize in the constructor with the failed precondition status, and then again when calling `Initialize()`. Then the only thing `GenerateVariants()` has to do is to `if (!initialize_status_.ok()) return initialize_status_;`

Also, once the initialization question is figured out, we need to have it carry its own state, like `current_variant_` and `wants_more_` so that it does ont modify anything.

But all of that, let's wait until we have the other pull requests done; we always can go back and make things 'prettier'.